miner/worker.go

```diff
@@ -391,3 +395,34 @@ func totalFees(block *types.Block, receipts []*types.Receipt) *big.Float {
 	}
 	return new(big.Float).Quo(new(big.Float).SetInt(feesWei), new(big.Float).SetInt(big.NewInt(params.Ether)))
 }
+
+// enforcePredicates takes a set of pending transactions (grouped by sender, and ordered by nonce)
+// and returns the subset of those transactions (following the same grouping) that satisfy predicateContext.
+// Any transaction that fails predicate verification will be removed from the tx pool and excluded
+// from the return value.
+// Transactions with a nonce that follows a removed transaction will be added back to the future
+// queue of the tx pool.
+func (w *worker) enforcePredicates(
+	rules params.Rules,
+	predicateContext *contract.PredicateContext,
+	pending map[common.Address]types.Transactions,
+) map[common.Address]types.Transactions {
+	result := make(map[common.Address]types.Transactions, len(pending))
+	for addr, txs := range pending {
+		for i, tx := range txs {
+			if err := core.CheckPredicates(rules, predicateContext, tx); err != nil {
+				log.Debug("Transaction predicate failed verification in miner", "sender", addr, "err", err)
+				// If the transaction fails the predicate check, we remove the transaction from the mempool
+				// and move all transactions from the same address with a subsequent nonce back to the
+				// future queue of the transaction pool.
+				w.eth.TxPool().RemoveTx(tx.Hash())
+				txs = txs[:i] // Cut off any transactions past the failed predicate in the return value
+				break
+			}
+		}
+		if len(txs) > 0 {
+			result[addr] = txs
+		}
```

Comment:
I believe a break would suffice if we use append-like approach for results. like
```go
      result[addr] = append(result[addr], tx)
```

Reply:
Do you mean putting this line within the for loop iterating transactions and calling append for each valid transaction?

Reply:
yep

Reply:
I'd prefer to do this once here